File: app/src/main/java/fr/free/nrw/commons/upload/UploadInterface.java
package fr.free.nrw.commons.upload;

import androidx.annotation.NonNull;
import com.google.gson.JsonObject;
import fr.free.nrw.commons.wikidata.WikidataConstants;
import io.reactivex.Observable;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;
import retrofit2.http.Field;
import retrofit2.http.FormUrlEncoded;
import retrofit2.http.Headers;
import retrofit2.http.Multipart;
import retrofit2.http.POST;
import retrofit2.http.Part;

public interface UploadInterface {

    @Multipart
    @POST(WikidataConstants.MW_API_PREFIX + "action=upload&stash=1&ignorewarnings=1")
    Observable<UploadResponse> uploadFileToStash(@Part("filename") RequestBody filename,
        @Part("filesize") RequestBody totalFileSize,
        @Part("offset") RequestBody offset,
        @Part("filekey") RequestBody fileKey,
        @Part("token") RequestBody token,
        @Part MultipartBody.Part filePart);

    @Headers("Cache-Control: no-cache")
    @POST(WikidataConstants.MW_API_PREFIX + "action=upload&ignorewarnings=1")
    @FormUrlEncoded
    @NonNull
    Observable<JsonObject> uploadFileFromStash(@NonNull @Field("token") String token,
        @NonNull @Field("text") String text,
        @NonNull @Field("comment") String comment,
        @NonNull @Field("filename") String filename,
        @NonNull @Field("filekey") String filekey);
}

File: app/src/main/java/fr/free/nrw/commons/upload/UploadInterface.kt
package fr.free.nrw.commons.upload

import com.google.gson.JsonObject
import io.reactivex.Observable
import okhttp3.MultipartBody
import okhttp3.RequestBody
import org.wikipedia.dataclient.Service
import retrofit2.http.*

interface UploadInterface {
    @Multipart
    @POST(Service.MW_API_PREFIX + "action=upload&stash=1&ignorewarnings=1")
    fun uploadFileToStash(
        @Part("filename") filename: RequestBody?,
        @Part("filesize") totalFileSize: RequestBody?,
        @Part("offset") offset: RequestBody?,
        @Part("filekey") fileKey: RequestBody?,
        @Part("token") token: RequestBody?,
        @Part filePart: MultipartBody.Part?
    ): Observable<UploadResponse>

    @Headers("Cache-Control: no-cache")
    @POST(Service.MW_API_PREFIX + "action=upload&ignorewarnings=1")
    @FormUrlEncoded
    fun uploadFileFromStash(
        @Field("token") token: String,
        @Field("text") text: String,
        @Field("comment") comment: String,
        @Field("filename") filename: String,
        @Field("filekey") filekey: String
    ): Observable<JsonObject>
}